Assets/Scripts/Utility/AssetInjection/WorldDataReplacement.cs

```diff
@@ -504,6 +513,83 @@ private static bool AssignBlockIndices(ref DFLocation dfLocation)
+        {
+            string fileName = GetDFBlockReplacementFilename(blockName, WorldDataVariants.NoVariant);
+
+            int jsonBlockIndex;
+            DFBlock? dfBlock = null; // Make blockData nullable
+            TextAsset blockReplacementJsonAsset;
+
+            // Seek from loose files
+            if (File.Exists(Path.Combine(worldDataPath, fileName)))
+            {
+                string blockReplacementJson = File.ReadAllText(Path.Combine(worldDataPath, fileName));
+                try
+                {
+                    dfBlock = (DFBlock)SaveLoadManager.Deserialize(typeof(DFBlock), blockReplacementJson);
+                }
+                catch (System.InvalidCastException e)
+                {
+                    Debug.LogError($"Invalid JSON format for block '{blockName}': {e.Message}");
+                    throw new WorldDataReplacementException(
+                        $"Block '{blockName}' JSON could not be cast to DFBlock.", e
+                    );
+                }
+            }
+            // Seek from mods
+            else if (ModManager.Instance != null && ModManager.Instance.TryGetAsset(fileName, false, out blockReplacementJsonAsset))
+            {
+                try
+                {
+                    dfBlock = (DFBlock)SaveLoadManager.Deserialize(typeof(DFBlock), blockReplacementJsonAsset.text);
+                }
+                catch (System.InvalidCastException e)
+                {
+                    Debug.LogError($"Invalid JSON format for block '{blockName}': {e.Message}");
+                    throw new WorldDataReplacementException(
+                        $"Block '{blockName}' JSON could not be cast to DFBlock.", e
+                    );
+                }
+            }
+
+            // Ensure blockData was successfully assigned
+            if (!dfBlock.HasValue)
+            {
+                Debug.LogError($"Failed to load block data for blockName: {blockName}");
+                throw new WorldDataReplacementException(
+                    $"Block '{blockName}' does not have a valid Index in its JSON file."
+                );
+            }
+
+            // Grab the variant key
+            string blockKey = blockName + WorldDataVariants.NoVariant;
+
+            // Check for the "Index" field and assign its value
+            jsonBlockIndex = dfBlock.Value.Index;
+
+            // If jsonBlockIndex is invalid (less than or equal to nextBlockIndex), use fallback method
+            if (jsonBlockIndex <= nextBlockIndex)
+            {
+                AssignNextIndex(blockName);
+
+                // Cache the full DFBlock
+                if (!blocks.ContainsKey(blockKey))
+                    blocks[blockKey] = dfBlock.Value;
+
+                return;
+            }
+
+            // Add to cache
+            newBlockNames[jsonBlockIndex] = blockName;
+            newBlockIndices[blockName] = jsonBlockIndex;
```

Comment:
Probably want to add a fallback to `AssignNextIndex` here too in case there are already entries for the key `jsonBlockIndex` and `blockName`. This way two different modded RMBs with the same custom index (by accident) can still get along.